### PR TITLE
Integer-valued polynomial ring

### LINKED ISSUE #34814: adding the ring of integer-valued polynomials
<div id="comment:0"></div>

in two different basis

CC:  @tscrim

Component: **commutative algebra**

Author: **Frédéric Chapoton**

_Issue created by migration from https://trac.sagemath.org/ticket/34814_

### PR DESCRIPTION
**Pull request by fchapoton**
Thanks for contributing to Sage! Detailed instructions to be added shortly.

Please make sure to also have a look at our
[Code Style Conventions](https://doc.sagemath.org/html/en/developer/coding_basics.html).

Fixes #34814

### CODE REVIEW

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 1]
```diff
@@ -0,0 +1,1144 @@
...
+
+                sage: F = PositiveBasis(ZZ)
+                sage: B = F.gen()
+                sage: (B+1).shift()
+                B[1] + B[2]
+            """
+            A = self.parent()
+            return A.sum(c * A.monomial(i + 1) for i, c in self)
+
+        def polynomial(self):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: Should we cache this?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 2]
```diff
@@ -0,0 +1,1144 @@
...
+            from sage.misc.sage_eval import sage_eval
+            return sage_eval(x, locals=self.gens_dict())
+        R = self.base_ring()
+        # coercion via base ring
+        x = R(x)
+        if x == 0:
+            return self.element_class(self, {})
+        return self.from_base_ring_from_one_basis(x)
+
+    def _coerce_impl(self, x):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: We aren't using the old coercion framework here, so is this actually called? Shouldn't this be put into `_coerce_map_from_` instead?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 3]
```diff
@@ -0,0 +1,1144 @@
...
+            B[5]
+            sage: x = polygen(QQ, 'x')
+            sage: A.from_polynomial(x)
+            -B[0] + B[1]
+        """
+        B = self.basis()
+        x = p.parent().gen()
+        remain = p
+        result = self.zero()
+        while remain != 0:
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
        while remain:
```

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 4]
```diff
@@ -0,0 +1,1144 @@
...
+            sage: A.product_on_basis(0, 1)
+            B[1]
+        """
+        i = ZZ(n1)
+        j = ZZ(n2)
+        if j < i:
+            j, i = i, j
+
+        return self.sum((-1)**k * i.binomial(k) * (i + j - k).binomial(i) *
+                        self.basis()[i + j - k] for k in range(i + 1))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I believe all of the terms of distinct since i < j. Thus, this can be simplified to
```suggestion
        R = self.base_ring()
        return self._from_dict({i+j-k: R((-1)**k * i.binomial(k) * (i + j - k).binomial(i)) for k in range(i + 1)})
```
If I am wrong, then make it `self.linear_combination()` to avoid the transient elements.

It might also be good to have a more detailed doctest here.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 5]
```diff
@@ -0,0 +1,1144 @@
...
+        Normalize input to ensure a unique representation.
+
+        EXAMPLES::
+
+            sage: F1 = IntegerValuedPolynomialRing(QQ)
+            sage: F2 = IntegerValuedPolynomialRing(QQ)
+            sage: F1 is F2
+            True
+        """
+        return super().__classcall__(cls, R)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This can be removed since it is effectively a no-op. The `__init__` will filter out extra inputs.

**Review by tscrim — CHANGES_REQUESTED** [review]
The same comments apply for the `PositiveBasis`. It seems like there would be a benefit to avoid some duplication through an ABC.

**Comment by mkoeppe** [comment]
https://github.com/sagemath/sage/issues/34814#issuecomment-1418191757 is unresolved

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 2 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: I have removed this and it works. Thanks for the suggestion.

**Comment by fchapoton** [comment]
Now in better state, working with realizations and only importing one thing into the global namespace

**Comment by tscrim** [comment]
@fchapoton Would you like me to go through what's there and leave some comments or are you still going to do some work on it?

**Comment by fchapoton** [comment]
Yes, please have a look if you can. I will not do anything else here for the next 7 days. Maybe moving some documentation to the top could be a good idea.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 6]
```diff
@@ -0,0 +1,1125 @@
...
+
+    TESTS::
+
+        sage: IntegerValuedPolynomialRing(24)
+        Traceback (most recent call last):
+        ...
+        TypeError: argument R must be a ring
+    """
+    def __init__(self, R):
+        if R not in Rings():
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: Probably should be `Rings().Commutative()` since this is going to be a commutative algebra.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 7]
```diff
@@ -0,0 +1,1125 @@
...
+            raise TypeError("argument R must be a ring")
+        self._base = R
+        cat = Algebras(R).Commutative().WithBasis()
+        Parent.__init__(self, base=R, category=cat.WithRealizations())
+
+        S = self.S()
+        B = self.B()
+        B.module_morphism(S._from_binomial_basis, codomain=S).register_as_coercion()
+        S.module_morphism(B._from_shifted_basis, codomain=B).register_as_coercion()
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This can lead to problems if `B` and `S` are garbage collected IIRC. This might be okay though as the realizations are stored and tied to the `IVPR`.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 8]
```diff
@@ -0,0 +1,1125 @@
...
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.a_realization()
+            Integer-Valued Polynomial Ring over Rational Field
+            in the shifted basis
+        """
+        return self.S()
+
+    def indices(self):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I don't think this is necessary.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 9]
```diff
@@ -0,0 +1,1125 @@
...
+
+        EXAMPLES::
+
+            sage: IntegerValuedPolynomialRing(QQ)
+            Integer-Valued Polynomial Ring over Rational Field
+        """
+        br = self.base_ring()
+        return f"Integer-Valued Polynomial Ring over {br}"
+
+    class Bases(Category_realization_of_parent):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:119]: I would rename this as `_Bases` as it is better to not be exposed under tab completion IMO.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 10]
```diff
@@ -0,0 +1,1125 @@
...
+                [Category of realizations of Integer-Valued Polynomial Ring
+                 over Rational Field,
+                 Join of Category of algebras with basis over Rational Field and
+                 Category of commutative algebras over Rational Field and
+                 Category of realizations of unital magmas]
+            """
+            A = self.base()
+            category = Algebras(A.base_ring()).Commutative()
+            return [A.Realizations(),
+                    category.Realizations().WithBasis()]
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
            category = Algebras(A.base_ring()).Commutative()
            return [A.Realizations(),
                    category.Realizations().WithBasis(),
                    category.WithBasis()]
```
Isn't there also a natural filtered algebra structure?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 11]
```diff
@@ -0,0 +1,1125 @@
...
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).B()
+                    sage: B = F.gen()
+                    sage: (B+1).shift()
+                    B[1] + B[2]
+                    sage: (B+1).shift(3)
+                    B[3] + B[4]
+                """
+                A = self.parent()
+                return A._from_dict({i + j: c for i, c in self})
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: You might want to make sure that the indices are in `self.basis().keys()` (which can be quickly accessed via `self._indices`). I could pass `1/2` or `4/2`, which might lead to subtle problems.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 12]
```diff
@@ -0,0 +1,1125 @@
...
+            \binom{x}{i} = \sum_{k=0}^{i} (-1)^{i-k} \binom{i}{k} \binom{x+k}{k}.
+
+
+        REFERENCES:
+
+        - :wikipedia:`Integer-valued polynomial`
+
+        INPUT:
+
+        - ``A`` -- a parent with realization
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I find this useless to confusing for a user since they will never directly construct this by calling the class on its own. So I would remove it.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 13]
```diff
@@ -0,0 +1,1125 @@
...
+
+        .. MATH::
+
+            \sum_{k=0}^{n_1} (-1)^k \binom{n_1}{k}\binom{n_1+n_2-k}{n_1} B[n_1 + n_2 - k].
+
+        There is a conversion formula between the two bases
+
+        .. MATH::
+
+            \binom{x}{i} = \sum_{k=0}^{i} (-1)^{i-k} \binom{i}{k} \binom{x+k}{k}.
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I am guessing a to-be-updated doc?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 14]
```diff
@@ -0,0 +1,1125 @@
...
+
+            EXAMPLES::
+
+                sage: F = IntegerValuedPolynomialRing(QQ).S(); F
+                Integer-Valued Polynomial Ring over Rational Field
+                in the shifted basis
+                sage: TestSuite(F).run()
+            """
+            CombinatorialFreeModule.__init__(self, A.base_ring(),
+                                             A.indices(),
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I would just make this a direct instance call. If you want to attach it to the abstract object, perhaps as an attribute?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 15]
```diff
@@ -0,0 +1,1125 @@
...
+                                             category=A.Bases(),
+                                             prefix="S",
+                                             latex_prefix=r"\mathbb{S}")
+
+        def _realization_name(self) -> str:
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:471]: This can be removed as it comes automatically from the `Sets().Realizations()` category.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 16]
```diff
@@ -0,0 +1,1125 @@
...
+            r"""
+            TESTS::
+
+                sage: F = IntegerValuedPolynomialRing(QQ).S()
+                sage: F._realization_name()
+                'shifted'
+            """
+            return "shifted"
+
+        # with @cached_method ?
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I wouldn't cache it as it is a "fast" type of multiplication.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 17]
```diff
@@ -0,0 +1,1125 @@
...
+                -2*S[2] + 3*S[3]
+            """
+            i = ZZ(n1)
+            j = ZZ(n2)
+            if j < i:
+                j, i = i, j
+
+            R = self.base_ring()
+            return self._from_dict({i + j - k: R((-1)**k * i.binomial(k) * (i + j - k).binomial(i))
+                                    for k in range(i + 1)})
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:504]:
```suggestion
            return self._from_dict({i + j - k: coeff for k in range(i + 1)
                                    if coeff := R((-1)**k * i.binomial(k) * (i + j - k).binomial(i)},
                                   remove_zeros=False)
```
I am so happy I found this (not so commonly used) syntax; it's great for these types of loops.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 18]
```diff
@@ -0,0 +1,1125 @@
...
+                sage: B = IntegerValuedPolynomialRing(ZZ).B()
+                sage: b = B.basis()
+                sage: S(b[3]+1)  # indirect doctest
+                3*S[1] - 3*S[2] + S[3]
+                sage: B(_)
+                B[0] + B[3]
+            """
+            i = ZZ(i)
+            R = self.base_ring()
+            return self._from_dict({k: R((-1)**(i - k) * i.binomial(k))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:526]: Similar to the previous comment.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 19]
```diff
@@ -0,0 +1,1125 @@
...
+
+        def from_polynomial(self, p):
+            """
+            Convert a polynomial into the ring of integer-valued polynomials.
+
+            This raises a ``ValueError`` if this is not possible.
+
+            INPUT:
+
+            - `p` -- a polynomial in one variable
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
            - ``p`` -- a polynomial in one variable
```

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 20]
```diff
@@ -0,0 +1,1125 @@
...
+                sage: ex = S[2]+S[4]
+                sage: A.from_h_vector(ex.h_vector())
+                S[2] + S[4]
+            """
+            B = self.basis()
+            d = len(h) - 1
+            m = matrix(QQ, d + 1, d + 1,
+                       lambda j, i: (-1)**(d - j) * binomial(d - i, d - j))
+            v = vector(QQ, [h[i] for i in range(d + 1)])
+            return self.sum(Integer(c) * B[i] for i, c in enumerate(m * v))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: You can use `self.linear_combination()`, which should be faster.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 21]
```diff
@@ -0,0 +1,1125 @@
...
+                sage: G._coerce_map_from_(QQ)
+                True
+                sage: F.has_coerce_map_from(PolynomialRing(ZZ,'x'))
+                False
+            """
+            # integer-valued polynomial rings over any base
+            # that coerces in:
+            if isinstance(R, IntegerValuedPolynomialRing.Shifted):
+                return self.base_ring().has_coerce_map_from(R.base_ring())
+            return self.base_ring().has_coerce_map_from(R)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This would also be a good place to do the coercion from the other basis (or bases should more come along).

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 22]
```diff
@@ -0,0 +1,1125 @@
...
+                    3/2
+
+                TESTS::
+
+                    sage: [(B**n).umbra() for n in range(1, 11)]
+                    [1/2, 1/6, 0, -1/30, 0, 1/42, 0, -1/30, 0, 5/66]
+                """
+                return self.shift().derivative_at_minus_one()
+
+            def delta(self):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:722]: Perhaps here we should slightly break with Python conventions since it is a capital Delta?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 23]
```diff
@@ -0,0 +1,1125 @@
...
+                    sage: S = A.basis()
+                    sage: S[5].variable_shift()
+                    S[0] + S[1] + S[2] + S[3] + S[4] + S[5]
+                """
+                if k == 0:
+                    return self
+
+                A = self.parent()
+                resu = A.sum(c * A.monomial(j) for i, c in self
+                             for j in range(i + 1))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This could likely be made faster by manually constructing the `dict`. You can also use `A.linear_combination()` for some speed too.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 24]
```diff
@@ -0,0 +1,1125 @@
...
+                    sage: S = A.basis()
+                    sage: S[5].variable_unshift()
+                    -S[4] + S[5]
+                """
+                if k == 0:
+                    return self
+
+                A = self.parent()
+                resu = (A.sum(c * A.monomial(i) for i, c in self) -
+                        A.sum(c * A.monomial(i - 1) for i, c in self if i))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: `A.linear_combination()` or `A._from_dict` for speed. Although isn't the first summation just `self`?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 25]
```diff
@@ -0,0 +1,1125 @@
...
+
+                EXAMPLES::
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).S()
+                    sage: B = F.gen()
+                    sage: (B+1).polynomial()
+                    x + 2
+                """
+                x = polygen(QQ, 'x')
+                return sum(c * binomial(x + i, i) for i, c in self)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This will have a different result for `0`. Also add a test for this corner case checking for a consistent parent.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 26]
```diff
@@ -0,0 +1,1125 @@
...
+                    sage: (B[2]*B[4]).sum()
+                    1
+                """
+                return sum(c for _, c in self)
+
+    S = Shifted
+
+    # =====     Another basis for the same algebra     =====
+
+    class Binomial(CombinatorialFreeModule, BindableClass):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:847]: Similar comments as for the shifted basis.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 8 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: could you please expand ? I do no understand. The method "indices" is used in each realization init.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 17 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:504]: I am not very fond of this walrus syntax. :)

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 22 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:722]: I would rather keep the little delta.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 8 above)
Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: You could just directly call `NonNegativeIntegers()` there instead. If you want to have some common place, then you could have an ABC for the bases, do something within the bases category, or make it a hidden attribute of the abstract object. To me, it also clutters up the abstract object’s namespace with something that is not so well-defined, and has other more “standard” ways to get access to it.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 15 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:471]: I do not understand. I have to tell the correct name somewhere, no ?

Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:471]: The `Sets().WithRealizations()` code pulls it automatically from the class name as a default.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 17 above)
Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:504]: Fair enough; this is likely a micro-optimization.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 10 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: indeed, done (hopefully)

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 8 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: ok, done

**Comment by fchapoton** [comment]
I wonder if I should try to factorise the "polynomial" and "from polynomial" methods ? I could create a `_basis_polynomial` method just returning the value of B[i] and S[i] as polynomials.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 9 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:119]: hmm, it should be exposed in the doc as it contains useful methods.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 11 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: done, not yet pushed

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 14 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: I do not understand, sorry

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 15 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:471]: but this would use the capitalized `Binomial` and `Shifted`

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 21 above)
Comment (fchapoton) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: better than the pair of `register_as_coercion` ?

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 14 above)
Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: What I mean is
```
            CombinatorialFreeModule.__init__(self, A.base_ring(),
                                             NonNegativeIntegers(),
```

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 15 above)
Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py:471]: Fair enough, although I don't think we are very consistent across Sage about this. Plus the code is already there too.

File: src/sage/rings/polynomial/integer_valued_polynomials.py (hunk 21 above)
Comment (tscrim) [reply src/sage/rings/polynomial/integer_valued_polynomials.py]: I think so. Both bases do not need to be created an nailed in memory (at creation of the abstract object), and it has one (standard) centralized place for setting the coercions. (It also means we don't have to worry about the initialization code creating the coercion maps `B -> S`  and `S -> B` becoming defunct because one of `B` or `S` gets garbage collected and then recreated, which then leads to further issues. This is something that has happened in the past.)

**Comment by tscrim** [comment]
> I wonder if I should try to factorise the "polynomial" and "from polynomial" methods ? I could create a `_basis_polynomial` method just returning the value of B[i] and S[i] as polynomials.

I think that would be good. It took me a bit to actually see what the difference between those methods actually ways.

**Comment by fchapoton** [comment]
> > I wonder if I should try to factorise the "polynomial" and "from polynomial" methods ? I could create a `_basis_polynomial` method just returning the value of B[i] and S[i] as polynomials.
> 
> I think that would be good. It took me a bit to actually see what the difference between those methods actually ways.

Ok, I have now done both the factorisation of `polynomial` and `from_polynomial` and the better coercion. This should start to look good. Is there anything that I missed ?

**Comment by tscrim** [comment]
I think you have taken care of everything. The only comment I have on the recent commits is that I think you can simply return the `module_morphism` in `_coerce_map_from_` so it is not recreated every time and that you know the base rings are compatible. Having it in the element constructor means it might pick up something going from, e.g., `Z/5Z` to `Z/2Z` as base rings.

**Comment by fchapoton** [comment]
> I think you have taken care of everything. The only comment I have on the recent commits is that I think you can simply return the `module_morphism` in `_coerce_map_from_` so it is not recreated every time and that you know the base rings are compatible. Having it in the element constructor means it might pick up something going from, e.g., `Z/5Z` to `Z/2Z` as base rings.

oh,  I see, thanks. I did not know that one could return a morphism and not a boolean in `_coerce_map_from_`. Done and it seems to work fine enough.

**Comment by tscrim** [comment]
> oh, I see, thanks. I did not know that one could return a morphism and not a boolean in `_coerce_map_from_`. Done and it seems to work fine enough.

It can also take an arbitrary Python callable object as well as I recall, which then gets wrapped into a coercion morphism.

Anyways, that's all my comments for now. Let me know when you are ready for a full review (which will probably be quite quick at this point).

**Comment by fchapoton** [comment]
ok, this should be almost ready.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 27]
```diff
@@ -0,0 +1,1106 @@
...
+    - :wikipedia:`Integer-valued polynomial`
+
+    TESTS::
+
+        sage: IntegerValuedPolynomialRing(24)
+        Traceback (most recent call last):
+        ...
+        TypeError: argument R must be a commutative ring
+    """
+    def __init__(self, R):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py:78]: This will need a doctest; the usual `TestSuite(IVPR).run()` should be good.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 28]
```diff
@@ -0,0 +1,1106 @@
...
+                        top_coeff = self.base_ring()(top_coeff)
+                    except TypeError as exc:
+                        msg = 'not a polynomial with integer'
+                        msg += f' values: {top_coeff}'
+                        raise ValueError(msg) from exc
+                    remain += -top_coeff * poly(N)
+                    result += top_coeff * B[N]
+                return result
+
+            def gen(self):
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: It would be good for this to take an (mostly ignored with a default of `0`) argument for consistency with other algebras/rings/etc.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 29]
```diff
@@ -0,0 +1,1106 @@
...
+                In the shifted basis, this is the evaluation at `x=0`.
+
+                EXAMPLES::
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).S()
+                    sage: B = F.basis()
+                    sage: (B[2]*B[4]).sum_of_coefficients()
+                    1
+                """
+                return sum(c for _, c in self)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
                return self.parent().base_ring().sum(self._monomial_coefficients.values())
```
The first part is needed for `0`, which would otherwise be an `int`.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 30]
```diff
@@ -0,0 +1,1106 @@
...
+
+                EXAMPLES::
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).S()
+                    sage: B = F.basis()
+                    sage: (3*B[4]+6*B[7]).content()
+                    3
+                """
+                from sage.arith.misc import gcd
+                return gcd(c for _, c in self)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
                return gcd(self._monomial_coefficients.values())
```
What should happen for `0`?

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 31]
```diff
@@ -0,0 +1,1106 @@
...
+                sage: S = A.basis()
+                sage: ex = S[2]+S[4]
+                sage: A.from_h_vector(ex.h_vector())
+                S[2] + S[4]
+            """
+            d = len(h) - 1
+            m = matrix(QQ, d + 1, d + 1,
+                       lambda j, i: (-1)**(d - j) * binomial(d - i, d - j))
+            v = vector(QQ, [h[i] for i in range(d + 1)])
+            return self._from_dict({i: Integer(c)
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
            R = self.base_ring()
            return self._from_dict({i: R(c)
```

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 32]
```diff
@@ -0,0 +1,1106 @@
...
+                sage: R = IntegerValuedPolynomialRing(QQ).S()
+                sage: x = R.gen()
+                sage: R(3)
+                3*S[0]
+                sage: R(x)
+                S[1]
+            """
+            if x in NonNegativeIntegers():
+                W = self.basis().keys()
+                return self.monomial(W(x))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: I am not sure about this. You get very different behaviors for `-2` and `2`. I feel like the output type should generally depend on the parent. However, I guess this should never be called whenever `ZZ` is a subring of the base ring as the coercion should get it first as per the doctest. So perhaps over a positive characteristic field this would be invoked...but it seems sketchy to me.

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 33]
```diff
@@ -0,0 +1,1106 @@
...
+
+            Elements of the integer-valued polynomial ring canonically
+            coerce in::
+
+                sage: x = F.gen()
+                sage: F.coerce(x*x) # indirect doctest
+                6*S[1] + 2*S[2]
+
+            Elements of the integers coerce in, since there is a coerce map
+            from `\ZZ` to GF(7)::
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]:
```suggestion
            from `\ZZ` to `\GF{7}`::
```

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 34]
```diff
@@ -0,0 +1,1106 @@
...
+
+                sage: G = IntegerValuedPolynomialRing(ZZ).S()
+                sage: Gx = G.gen()
+                sage: z = F.coerce(Gx**2); z
+                -S[1] + 2*S[2]
+                sage: z.parent() is F
+                True
+
+            However, `\GF{7}` does not coerce to `\ZZ`, so the shuffle
+            algebra over `\GF{7}` does not coerce to the one over `\ZZ`::
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: This is not the shuffle algebra. ;)

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 35]
```diff
@@ -0,0 +1,1106 @@
...
+                sage: R = IntegerValuedPolynomialRing(QQ).B()
+                sage: x = R.gen()
+                sage: R(3)
+                3*B[0]
+                sage: R(x)
+                B[1]
+            """
+            if x in NonNegativeIntegers():
+                W = self.basis().keys()
+                return self.monomial(W(x))
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: Same comment as the other `_element_constructor_`

File: src/sage/rings/polynomial/integer_valued_polynomials.py [hunk 36]
```diff
@@ -0,0 +1,1106 @@
...
+
+                sage: G = IntegerValuedPolynomialRing(ZZ).B()
+                sage: Gx = G.gen()
+                sage: z = F.coerce(Gx**2); z
+                B[1] + 2*B[2]
+                sage: z.parent() is F
+                True
+
+            However, `\GF{7}` does not coerce to `\ZZ`, so the shuffle
+            algebra over `\GF{7}` does not coerce to the one over `\ZZ`::
```
Comment (tscrim) [inline src/sage/rings/polynomial/integer_valued_polynomials.py]: Still not the shuffle algebra. :p

**Review by tscrim — CHANGES_REQUESTED** [review]
Just a few more little things.

**Review by tscrim — APPROVED** [review]
Thank you. LGTM.
